File: app/src/main/java/com/rosasoft/wintalker/engine/Transcriber.java
package com.rosasoft.wintalker.engine;

import java.util.ArrayList;
import java.util.List;

/**
 * Lithuanian grapheme→phoneme transcriber — pure Java port of the verified C
 * reimplementation (engine/transcr_c), which matches the original libtranscr.so
 * 100% over 515 words. No native code, so it runs on any ABI incl. arm64-v8a.
 *
 * Input is a cp1257-uppercased word (see normalise()); output is the phoneme
 * token list the synthesiser consumes, with '_' word boundaries.
 */
public final class Transcriber {

    // cp1257 Lithuanian letters used as constants
    private static final int A_OG = 0xc0, C_HK = 0xc8, E_OG = 0xc6, E_DOT = 0xcb,
            I_OG = 0xc1, S_HK = 0xd0, U_OG = 0xd8, U_MAC = 0xdb, Z_HK = 0xde;

    private static boolean isFrontVowel(int c) {
        return c == 'I' || c == 'Y' || c == I_OG || c == 'E' || c == E_DOT || c == E_OG;
    }

    private static String consVoiced(int c) {
        switch (c) {
            case 'B': return "b"; case 'C': return "ts"; case 'D': return "d";
            case 'F': return "f"; case 'G': return "g"; case 'H': return "h";
            case 'K': return "k"; case 'L': return "l"; case 'M': return "m";
            case 'N': return "n"; case 'P': return "p"; case 'R': return "r";
            case 'S': return "s"; case 'T': return "t"; case 'V': return "v";
            case 'Z': return "z"; case 'J': return "j'";
            case C_HK: return "tS"; case S_HK: return "S"; case Z_HK: return "Z";
            default: return null;
        }
    }

    private static String consDevoiced(int c) {
        switch (c) {
            case 'B': return "p"; case 'D': return "t"; case 'G': return "k";
            case 'Z': return "s"; case Z_HK: return "S"; case S_HK: return "S";
            case 'C': return "ts"; case C_HK: return "tS";
            default: return consVoiced(c);
        }
    }

    private static String vowelPhoneme(int c) {
        switch (c) {
            case 'A': case A_OG: return "aA";
            case 'E': case E_OG: return "eA";
            case E_DOT: return "eE";
            case 'I': return "i"; case I_OG: return "iI"; case 'Y': return "iI";
            case 'O': return "oO";
            case 'U': return "u"; case U_MAC: return "uU"; case U_OG: return "uU";
            default: return null;
        }
    }

    /** Regressive palatalisation through palatalised consonant chains. */
    private static boolean isPalatalised(int[] w, int n, int i) {
        if (i + 1 >= n) return false;
        int nx = w[i + 1];
        if (isFrontVowel(nx) || nx == 'I' || nx == 'J') return true;
        if (vowelPhoneme(nx) != null) return false;
        if (consVoiced(nx) != null) return isPalatalised(w, n, i + 1);
        return false;
    }

    /** Transcribe a cp1257-uppercased word into phoneme tokens incl. '_' bounds. */
    public static List<String> transcribe(int[] w, int n) {
        List<String> out = new ArrayList<>();
        out.add("_");
        for (int i = 0; i < n; i++) {
            int c = w[i];

            // CH → x
            if (c == 'C' && i + 1 < n && w[i + 1] == 'H') {
                out.add(isPalatalised(w, n, i + 1) ? "x'" : "x");
                i++; continue;
            }
            // DŽ → dZ, DZ → dz
            if (c == 'D' && i + 1 < n && (w[i + 1] == Z_HK || w[i + 1] == 'Z')) {
                String base = (w[i + 1] == Z_HK) ? "dZ" : "dz";
                out.add(isPalatalised(w, n, i + 1) ? base + "'" : base);
                i++; continue;
            }
            // softness 'i' before a back vowel is absorbed
            if (c == 'I' && !(i > 0 && vowelPhoneme(w[i - 1]) != null) && i + 1 < n) {
                int nx = w[i + 1];
                boolean back = (nx == 'A' || nx == A_OG || nx == 'O' || nx == 'U'
                        || nx == U_OG || nx == U_MAC);
                if (back) continue;
            }
            // diphthong glides after short vowels
            if (i > 0) {
                int pv = w[i - 1];
                boolean shortV = (pv == 'A' || pv == 'E' || pv == 'O' || pv == 'U');
                if (shortV) {
                    if (c == 'I') { out.add("J"); continue; }
                    if (c == 'U' && pv != 'U') { out.add("W"); continue; }
                }
            }

            String vp = vowelPhoneme(c);
            if (vp != null) {
                if ((c == 'A' || c == A_OG) && i > 0
                        && ((w[i - 1] == 'I' && !(i >= 2 && vowelPhoneme(w[i - 2]) != null))
                            || w[i - 1] == 'J')) {
                    out.add("eA"); continue;
                }
                out.add(vp); continue;
            }

            // sonorant uppercase positional variants
            {
                boolean nextIsCons = (i + 1 >= n) || (vowelPhoneme(w[i + 1]) == null && w[i + 1] != 'I');
                boolean prevIsCons = (i > 0) && (vowelPhoneme(w[i - 1]) == null);
                boolean afterVowel = (i > 0) && vowelPhoneme(w[i - 1]) != null;
                boolean coda = afterVowel && nextIsCons;
                boolean prevObstruent = prevIsCons
                        && !(w[i-1]=='L'||w[i-1]=='M'||w[i-1]=='N'||w[i-1]=='R'||w[i-1]=='J'||w[i-1]=='V');
                boolean pal = isPalatalised(w, n, i);
                boolean sonUp = coda || prevObstruent;
                boolean rUp = coda || prevIsCons || (nextIsCons && i + 1 < n && consVoiced(w[i+1]) != null);
                String up = null;
                if (c == 'L' && sonUp) up = "L";
                else if (c == 'M' && sonUp) up = "M";
                else if (c == 'N' && sonUp) up = "N";
                else if (c == 'R' && rUp) up = "R";
                if (up != null) { out.add(pal ? up + "'" : up); continue; }
            }

            String cp = consVoiced(c);
            if (cp == null) continue;
            boolean isFinal = (i == n - 1);
            String use = cp;
            if (isFinal) {
                use = consDevoiced(c);
            } else {
                int nx = w[i + 1];
                boolean nxUnvoiced = (nx=='P'||nx=='T'||nx=='K'||nx=='S'||nx==S_HK||nx=='C'||nx==C_HK||nx=='F'||nx=='H');
                boolean nxVoiced = (nx=='B'||nx=='D'||nx=='G'||nx=='Z'||nx==Z_HK);
                if (nxUnvoiced) {
                    use = consDevoiced(c);
                } else if (nxVoiced) {
                    switch (c) {
                        case 'S': use = "z"; break;
                        case S_HK: use = "Z"; break;
                        case 'K': use = "g"; break;
                        case 'P': use = "b"; break;
                        case 'T': use = "d"; break;
                        case 'C': use = "dz"; break;
                        case C_HK: use = "dZ"; break;
                        default: use = cp; break;
                    }
                }
            }
            boolean palatal = isPalatalised(w, n, i);
            if (palatal && !use.endsWith("'")) use = use + "'";
            out.add(use);
        }
        out.add("_");
        return out;
    }

    /** Uppercase a Unicode Lithuanian string to a cp1257 code array for transcribe(). */
    public static int[] normalise(String text) {
        int[] w = new int[text.length()];
        int n = 0;
        for (int i = 0; i < text.length(); i++) {
            char ch = Character.toUpperCase(text.charAt(i));
            int cp;
            switch (ch) {
                case 'Ą': cp = A_OG; break;  // Ą
                case 'Č': cp = C_HK; break;  // Č
                case 'Ę': cp = E_OG; break;  // Ę
                case 'Ė': cp = E_DOT; break; // Ė
                case 'Į': cp = I_OG; break;  // Į
                case 'Š': cp = S_HK; break;  // Š
                case 'Ų': cp = U_OG; break;  // Ų
                case 'Ū': cp = U_MAC; break; // Ū
                case 'Ž': cp = Z_HK; break;  // Ž
                default:
                    if (ch >= 'A' && ch <= 'Z') cp = ch;
                    else if (ch == ' ' || ch == '\t') continue;
                    else cp = -1;
            }
            if (cp > 0) w[n++] = cp;
        }
        int[] r = new int[n];
        System.arraycopy(w, 0, r, 0, n);
        return r;
    }
}

File: core/src/transcriber.rs
//! Lithuanian grapheme→phoneme transcriber — faithful Rust port of the Java
//! `Transcriber` (which matches the original libtranscr.so 100% over 515 words).
//! Input is a cp1257-uppercased word (see `normalise`); output is the phoneme
//! token list with `_` word boundaries.

// cp1257 Lithuanian letters used as constants
const A_OG: i32 = 0xc0;
const C_HK: i32 = 0xc8;
const E_OG: i32 = 0xc6;
const E_DOT: i32 = 0xcb;
const I_OG: i32 = 0xc1;
const S_HK: i32 = 0xd0;
const U_OG: i32 = 0xd8;
const U_MAC: i32 = 0xdb;
const Z_HK: i32 = 0xde;

const A: i32 = b'A' as i32;
const Z: i32 = b'Z' as i32;

#[inline]
fn is_front_vowel(c: i32) -> bool {
    c == b'I' as i32 || c == b'Y' as i32 || c == I_OG || c == b'E' as i32 || c == E_DOT || c == E_OG
}

fn cons_voiced(c: i32) -> Option<&'static str> {
    Some(match c as u8 as char {
        'B' => "b", 'C' => "ts", 'D' => "d", 'F' => "f", 'G' => "g", 'H' => "h",
        'K' => "k", 'L' => "l", 'M' => "m", 'N' => "n", 'P' => "p", 'R' => "r",
        'S' => "s", 'T' => "t", 'V' => "v", 'Z' => "z", 'J' => "j'",
        _ => match c {
            C_HK => "tS",
            S_HK => "S",
            Z_HK => "Z",
            _ => return None,
        },
    })
}

fn cons_devoiced(c: i32) -> Option<&'static str> {
    match c as u8 as char {
        'B' => Some("p"), 'D' => Some("t"), 'G' => Some("k"), 'Z' => Some("s"),
        'C' => Some("ts"),
        _ => match c {
            Z_HK => Some("S"),
            S_HK => Some("S"),
            C_HK => Some("tS"),
            _ => cons_voiced(c),
        },
    }
}

fn vowel_phoneme(c: i32) -> Option<&'static str> {
    match c as u8 as char {
        'A' => Some("aA"),
        'E' => Some("eA"),
        'I' => Some("i"),
        'Y' => Some("iI"),
        'O' => Some("oO"),
        'U' => Some("u"),
        _ => match c {
            A_OG => Some("aA"),
            E_OG => Some("eA"),
            E_DOT => Some("eE"),
            I_OG => Some("iI"),
            U_MAC => Some("uU"),
            U_OG => Some("uU"),
            _ => None,
        },
    }
}

/// Regressive palatalisation through palatalised consonant chains.
fn is_palatalised(w: &[i32], n: usize, i: usize) -> bool {
    if i + 1 >= n {
        return false;
    }
    let nx = w[i + 1];
    if is_front_vowel(nx) || nx == b'I' as i32 || nx == b'J' as i32 {
        return true;
    }
    if vowel_phoneme(nx).is_some() {
        return false;
    }
    if cons_voiced(nx).is_some() {
        return is_palatalised(w, n, i + 1);
    }
    false
}

/// Transcribe a cp1257-uppercased word into phoneme tokens incl. `_` bounds.
pub fn transcribe(w: &[i32], n: usize) -> Vec<String> {
    let mut out: Vec<String> = Vec::new();
    out.push("_".to_string());
    let mut i = 0usize;
    while i < n {
        let c = w[i];

        // CH -> x
        if c == b'C' as i32 && i + 1 < n && w[i + 1] == b'H' as i32 {
            out.push(if is_palatalised(w, n, i + 1) { "x'" } else { "x" }.to_string());
            i += 2;
            continue;
        }
        // DŽ -> dZ, DZ -> dz
        if c == b'D' as i32 && i + 1 < n && (w[i + 1] == Z_HK || w[i + 1] == b'Z' as i32) {
            let base = if w[i + 1] == Z_HK { "dZ" } else { "dz" };
            out.push(if is_palatalised(w, n, i + 1) { format!("{}'", base) } else { base.to_string() });
            i += 2;
            continue;
        }
        // softness 'i' before a back vowel is absorbed
        if c == b'I' as i32
            && !(i > 0 && vowel_phoneme(w[i - 1]).is_some())
            && i + 1 < n
        {
            let nx = w[i + 1];
            let back = nx == b'A' as i32 || nx == A_OG || nx == b'O' as i32 || nx == b'U' as i32
                || nx == U_OG || nx == U_MAC;
            if back {
                i += 1;
                continue;
            }
        }
        // diphthong glides after short vowels
        if i > 0 {
            let pv = w[i - 1];
            let short_v = pv == b'A' as i32 || pv == b'E' as i32 || pv == b'O' as i32 || pv == b'U' as i32;
            if short_v {
                if c == b'I' as i32 {
                    out.push("J".to_string());
                    i += 1;
                    continue;
                }
                if c == b'U' as i32 && pv != b'U' as i32 {
                    out.push("W".to_string());
                    i += 1;
                    continue;
                }
            }
        }

        if let Some(vp) = vowel_phoneme(c) {
            if (c == b'A' as i32 || c == A_OG)
                && i > 0
                && ((w[i - 1] == b'I' as i32 && !(i >= 2 && vowel_phoneme(w[i - 2]).is_some()))
                    || w[i - 1] == b'J' as i32)
            {
                out.push("eA".to_string());
                i += 1;
                continue;
            }
            out.push(vp.to_string());
            i += 1;
            continue;
        }

        // sonorant uppercase positional variants
        {
            let next_is_cons = (i + 1 >= n) || (vowel_phoneme(w[i + 1]).is_none() && w[i + 1] != b'I' as i32);
            let prev_is_cons = i > 0 && vowel_phoneme(w[i - 1]).is_none();
            let after_vowel = i > 0 && vowel_phoneme(w[i - 1]).is_some();
            let coda = after_vowel && next_is_cons;
            let prev_obstruent = prev_is_cons
                && !(w[i - 1] == b'L' as i32 || w[i - 1] == b'M' as i32 || w[i - 1] == b'N' as i32
                    || w[i - 1] == b'R' as i32 || w[i - 1] == b'J' as i32 || w[i - 1] == b'V' as i32);
            let pal = is_palatalised(w, n, i);
            let son_up = coda || prev_obstruent;
            let r_up = coda || prev_is_cons
                || (next_is_cons && i + 1 < n && cons_voiced(w[i + 1]).is_some());
            let up = if c == b'L' as i32 && son_up {
                Some("L")
            } else if c == b'M' as i32 && son_up {
                Some("M")
            } else if c == b'N' as i32 && son_up {
                Some("N")
            } else if c == b'R' as i32 && r_up {
                Some("R")
            } else {
                None
            };
            if let Some(up) = up {
                out.push(if pal { format!("{}'", up) } else { up.to_string() });
                i += 1;
                continue;
            }
        }

        let cp = match cons_voiced(c) {
            Some(s) => s,
            None => {
                i += 1;
                continue;
            }
        };
        let is_final = i == n - 1;
        let mut use_: String = cp.to_string();
        if is_final {
            use_ = cons_devoiced(c).unwrap_or(cp).to_string();
        } else {
            let nx = w[i + 1];
            let nx_unvoiced = nx == b'P' as i32 || nx == b'T' as i32 || nx == b'K' as i32
                || nx == b'S' as i32 || nx == S_HK || nx == b'C' as i32 || nx == C_HK
                || nx == b'F' as i32 || nx == b'H' as i32;
            let nx_voiced = nx == b'B' as i32 || nx == b'D' as i32 || nx == b'G' as i32
                || nx == b'Z' as i32 || nx == Z_HK;
            if nx_unvoiced {
                use_ = cons_devoiced(c).unwrap_or(cp).to_string();
            } else if nx_voiced {
                use_ = match c as u8 as char {
                    'S' => "z".to_string(),
                    'K' => "g".to_string(),
                    'P' => "b".to_string(),
                    'T' => "d".to_string(),
                    'C' => "dz".to_string(),
                    _ => match c {
                        S_HK => "Z".to_string(),
                        C_HK => "dZ".to_string(),
                        _ => cp.to_string(),
                    },
                };
            }
        }
        let palatal = is_palatalised(w, n, i);
        if palatal && !use_.ends_with('\'') {
            use_.push('\'');
        }
        out.push(use_);
        i += 1;
    }
    out.push("_".to_string());
    out
}

/// Uppercase a Unicode Lithuanian string to a cp1257 code array for `transcribe`.
pub fn normalise(text: &str) -> Vec<i32> {
    let mut w: Vec<i32> = Vec::with_capacity(text.len());
    for ch0 in text.chars() {
        // Uppercase (Lithuanian ą->Ą, č->Č, ... are 1:1 in Unicode).
        let ch = ch0.to_uppercase().next().unwrap_or(ch0);
        let cp: i32 = match ch {
            'Ą' => A_OG,
            'Č' => C_HK,
            'Ę' => E_OG,
            'Ė' => E_DOT,
            'Į' => I_OG,
            'Š' => S_HK,
            'Ų' => U_OG,
            'Ū' => U_MAC,
            'Ž' => Z_HK,
            ' ' | '\t' => continue,
            _ => {
                let u = ch as i32;
                if u >= A && u <= Z {
                    u
                } else {
                    -1
                }
            }
        };
        if cp > 0 {
            w.push(cp);
        }
    }
    w
}
